packages/flutter_tools/lib/src/device.dart

```diff
@@ -112,14 +112,48 @@ abstract class DeviceManager {
         device.id.toLowerCase().startsWith(deviceId) ||
         device.name.toLowerCase().startsWith(deviceId);
 
-    final Device exactMatch = devices.firstWhere(
-        exactlyMatchesDeviceId, orElse: () => null);
-    if (exactMatch != null) {
-      return <Device>[exactMatch];
+    // Some discoverers have hard-coded device IDs and return quickly, and others
+    // shell out to other processes and can take longer.
+    // Process discoverers as they can return results, so if an exact match is
+    // found quickly, we don't wait for all the discoverers to complete.
+    final List<Device> prefixMatches = <Device>[];
+    final Completer<Device> exactMatchCompleter = Completer<Device>();
+    final List<StreamSubscription<List<Device>>> futureDevices = <StreamSubscription<List<Device>>>[];
+    for (final DeviceDiscovery discoverer in _platformDiscoverers) {
+      futureDevices.add(
+        discoverer
+        .devices
+        .asStream()
+        .listen((List<Device> devices) {
```

Comment:
I'd suggest investigating whether it makes sense to pass an `onError` callback to this `listen()`. If there is an error added to a `discoverer` stream, you probably don't want it to propagate asynchronously to the containing zone.

Reply:
`discoverer.devices` is a `Future<List<Device>>`, I'm making it into a stream with `.asStream()`.  Are there cases there could be errors on that stream?  Streams give me heartburn.

Reply:
I just realized this doesn't need to be a stream at all.